docs/proposals/20230726-integrate-keda-http-add-on.md

```diff
@@ -0,0 +1,285 @@
+  3. KEDA scales the target (Function) that reaches the threshold value according to the scaling options.
+- **Inject Dapr sidecar**:
+  1. Dapr control plane includes the [dapr-sidecar-injector service](https://docs.dapr.io/operations/hosting/kubernetes/kubernetes-overview/), which watches for new pods with the `dapr.io/enabled` annotation and injects a container with the `daprd` process within the pod.
+  2. In this case, sidecar arguments can be passed through annotations as outlined in the **Kubernetes annotations** column in [this table](https://docs.dapr.io/reference/arguments-annotations-overview/).
+
+![sync runtime overview](../images/integrate_keda_http-add-on.png)
+
+![KEDA+Dapr architecture overview](../images/function-keda-dapr.png)
+
+### CRD
+
+Add an `HTTPScaledObject ` type in `OpenFunction/apis/core/v1beta2/serving_types.go`:
+
+```go
+type HTTPScaledObject struct {
+	// Target metric value
+	// +optional
+	TargetPendingRequests *int32 `json:"targetPendingRequests,omitempty"`
+	// Cooldown period value
+	// +optional
+	CooldownPeriod *int32 `json:"cooldownPeriod,omitempty"`
+}
+
+type KedaScaleOptions struct {
+	// +optional
+	ScaledObject *KedaScaledObject `json:"scaledObject,omitempty"`
+	// +optional
+	HTTPScaledObject *HTTPScaledObject `json:"httpScaledObject,omitempty"`
+	// +optional
+	ScaledJob *KedaScaledJob `json:"scaledJob,omitempty"`
+	// Triggers are used to specify the trigger sources of the function.
+	// The Keda (ScaledObject, ScaledJob) configuration in ScaleOptions cannot take effect without Triggers being set.
+	// +optional
+	Triggers []kedav1alpha1.ScaleTriggers `json:"triggers,omitempty"`
+}
+```
+
+The `Function` yaml will be like this:
+
+```yaml
+apiVersion: core.openfunction.io/v1beta2
+kind: Function
+metadata:
+  name: logs-async-handler
+spec:
+  version: "v2.0.0"
+  image: openfunctiondev/logs-async-handler:v1
+  imageCredentials:
+    name: push-secret
+  build:
+    builder: openfunction/builder-go:latest
+    env:
+      FUNC_NAME: "LogsHandler"
+      FUNC_CLEAR_SOURCE: "true"
+#     # Use FUNC_GOPROXY to set the goproxy if failed to fetch go modules
+#     FUNC_GOPROXY: "https://goproxy.cn"
+    srcRepo:
+      url: "https://github.com/OpenFunction/samples.git"
+      sourceSubPath: "functions/async/logs-handler-function/"
+      revision: "main"
+  serving:
+    triggers:
+      http:
+        port: "8080"
+        route:
+          hostnames:
+            - "async-demo.com"
+          rules:
+            - matches:
+                - path:
+                    type: PathPrefix
+                    value: /
+              backendRefs:
+                - name: async-demo-svc
+                  port: 8080
+        inputs:
+          - type: binding.kafka 
+            name: kafka-input
+          - type: pubsub.rocketmq
+            name: rocketmq-input
+            topic: sample
+          - type: state.redis
+            name: redis-input 
+      dapr:
```

Comment:
so with keda-http, we can set http trigger together with dapr trigger? @tpiperatgod ?

Reply:
I don't think so. A function can only have http trigger or dapr trigger, because the EDA's engine can only be set to keda-http-addon or keda scaler, at least for now.

Reply:
> I don't think so. A function can only have http trigger or dapr trigger, because the EDA's engine can only be set to keda-http-addon or keda scaler, at least for now.

Then we should change this example to limit to the keda-http sync function with HTTP Trigger. No need to add dapr trigger for async stuff.

And the function name and svc can be changed from async-xxx to sync-xxx